### PR TITLE
Add new drive code

### PR DESCRIPTION
**Pull request by mpatankar6**
[drivetrain-improvements.pdf](https://github.com/team4909/2024-Crescendo/files/14076018/drivetrain-improvements.pdf)

### CODE REVIEW

File: .github/workflows/build.yml
Comment (TechplexEngineer) [inline .github/workflows/build.yml:1]: Nice!

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=permwrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.5-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.4-bin.zip
```
Comment (TechplexEngineer) [inline gradle/wrapper/gradle-wrapper.properties]: Do you know if there is a reason for the gradle downgrade?

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -1,136 +1,204 @@
...
+        new RepeatCommand(m_intake.intake()),
+        new RepeatCommand(m_shooter.Intake())
+            .until(
+                () -> {
+                  return mytimeofflight.getRange()
+                      <= SmartDashboard.getNumber("StopDistance", defaultStopDistance);
+                }));
   }
 
-  /**
```
Comment (TechplexEngineer) [inline src/main/java/frc/robot/Robot.java:55]: Can we add these comments back? They are helpful for the new team members.

File: src/main/java/frc/robot/drivetrain/Drivetrain.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+import java.util.function.Consumer;
+import java.util.function.DoubleSupplier;
+import java.util.function.UnaryOperator;
+import org.littletonrobotics.junction.AutoLogOutput;
+import org.littletonrobotics.junction.Logger;
+
+public class Drivetrain extends SubsystemBase {
+
+  public final double kTrackwidthMeters = Units.inchesToMeters(26.0);
+  public final double kWheelbaseMeters = Units.inchesToMeters(26.0);
```
Comment (TechplexEngineer) [inline src/main/java/frc/robot/drivetrain/Drivetrain.java:42]: Should this be 16" which is the distance from the front wheel to the back wheel

File: src/main/java/frc/robot/drivetrain/ImuIO.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package frc.robot.drivetrain;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import org.littletonrobotics.junction.AutoLog;
+
+public interface ImuIO {
```
Comment (TechplexEngineer) [inline src/main/java/frc/robot/drivetrain/ImuIO.java:6]: It is unclear to me how this class is used. Can you add a comment or provide a link?

File: src/main/java/frc/robot/drivetrain/Module.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModulePosition;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.math.util.Units;
+import org.littletonrobotics.junction.Logger;
+
+public class Module {
+
+  public static final double kDriveRatio = 6.25;
```
Comment (TechplexEngineer) [inline src/main/java/frc/robot/drivetrain/Module.java]: Can you add a comment describing where this value came from? My guess was from the SDS module but their listed value is 6.12

image

File: src/main/java/frc/robot/drivetrain/Module.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  public static final double kSteerRatio = 150.0 / 7.0;
+  public final double kWheelDiameterMeters = Units.inchesToMeters(4.0);
+  public final double kWheelRadiusMeters = kWheelDiameterMeters / 2.0;
+  private final double kCouplingGearRatio = 50.0 / 14.0;
```
Comment (TechplexEngineer) [inline src/main/java/frc/robot/drivetrain/Module.java:16]: This value does not seem to be used.

File: src/main/java/frc/robot/drivetrain/ImuIO.java (hunk 4 above)
Comment (mpatankar6) [reply src/main/java/frc/robot/drivetrain/ImuIO.java:6]: The akit docs describe how IO layers work

https://github.com/Mechanical-Advantage/AdvantageKit/blob/main/docs/DATA-FLOW.md
https://github.com/Mechanical-Advantage/AdvantageKit/blob/main/docs/RECORDING-INPUTS.md

**Review by TechplexEngineer — APPROVED** [review]

**Pull request merged**